protoarray/src/main/java/tech/pegasys/teku/protoarray/ForkChoiceStrategy.java

```diff
@@ -13,21 +13,379 @@
 
 package tech.pegasys.teku.protoarray;
 
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
 import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.datastructures.blocks.BlockAndCheckpointEpochs;
+import tech.pegasys.teku.spec.datastructures.blocks.StateAndBlockSummary;
 import tech.pegasys.teku.spec.datastructures.forkchoice.ReadOnlyForkChoiceStrategy;
+import tech.pegasys.teku.spec.datastructures.forkchoice.ReadOnlyStore;
+import tech.pegasys.teku.spec.datastructures.forkchoice.VoteTracker;
 import tech.pegasys.teku.spec.datastructures.forkchoice.VoteUpdater;
 import tech.pegasys.teku.spec.datastructures.operations.IndexedAttestation;
 import tech.pegasys.teku.spec.datastructures.state.Checkpoint;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.BeaconState;
+import tech.pegasys.teku.util.config.Constants;
 
-public interface ForkChoiceStrategy extends ReadOnlyForkChoiceStrategy {
+public class ForkChoiceStrategy implements BlockMetadataStore, ReadOnlyForkChoiceStrategy {
+  private static final Logger LOG = LogManager.getLogger();
+  private final ReadWriteLock protoArrayLock = new ReentrantReadWriteLock();
+  private final ReadWriteLock votesLock = new ReentrantReadWriteLock();
+  private final ReadWriteLock balancesLock = new ReentrantReadWriteLock();
+  private final ProtoArray protoArray;
+
+  private List<UInt64> balances;
+
+  private ForkChoiceStrategy(ProtoArray protoArray, List<UInt64> balances) {
+    this.protoArray = protoArray;
+    this.balances = balances;
+  }
+
+  // Public
+  public static SafeFuture<ForkChoiceStrategy> initializeAndMigrateStorage(
+      ReadOnlyStore store, ProtoArrayStorageChannel storageChannel) {
+    LOG.info("Migrating protoarray storing from snapshot to block based");
```

Comment:
do we need this as an info?

Reply:
I think its worth noting - the migration is a one off process so its not overly noisy.

For the record, this wasn't new in this PR - functionality is unchanged here, just shuffling class names around.